### PR TITLE
fix: tolerate streamed final text artifacts

### LINKED ISSUE #53449: Telegram: non-overflow replies duplicated when stream consumer's final_* delivery flags are lost (cancel/cache-invalidation); exact visible-vs-final match too strict
### Summary

On Telegram, **short/medium** assistant replies (well under the 4096 limit — observed 182–678 chars) are sometimes delivered **twice as two identical messages**: the streamed copy stays on screen, then the gateway's normal `send()` fires a second identical copy.

This is **not** the >4096 overflow/split duplication already covered by #51010 / #50965 / #52095 (no chunking involved here), and not the llama.cpp raw+formatted render of #53179. The trigger is the end-of-run **duplicate-suppression check** in `gateway/run.py` being too strict when the stream consumer's delivery flags get cleared while its text is already visible.

### Environment

- Hermes Agent v0.17.0, upstream commit `65be0061` (observed on a downstream fork, but the relevant logic is upstream).
- Platform: Telegram (DM, non-overflow replies). Streaming enabled.

### Root cause

At the end of `_run_agent`, the gateway decides whether to suppress the normal `send()` because streaming already delivered the answer. The decision reads the stream consumer (`_sc`):

- `_streamed = _sc.final_response_sent`
- `_content_delivered = _sc.final_content_delivered`
- fallback `_visible_matches_final = (_sc._visible_prefix() == _sc._clean_for_display(final))` — **exact** equality.

In the duplicate cases, the consumer has the final answer **on screen** (`_visible_prefix()` / `_last_sent_text` is non-empty) but **all** delivery flags are at their defaults: `final_response_sent=False`, `final_content_delivered=False`, `message_id=None`, `already_sent=False`. So `_streamed`/`_content_delivered` are both False, and the **exact** `_visible_matches_final` comparison also fails because the on-screen text differs from the final by trailing cursor/markdown/blank-line artifacts. Suppression is skipped → normal `send()` → duplicate.

Two ways the flags end up cleared while the text remains visible:

1. **Stream-task cancel race.** The `finally` block waits `asyncio.wait_for(stream_task, timeout=5.0)` and cancels on timeout. The `CancelledError` handler only confirms delivery `if self._accumulated and self._message_id:` — but `_message_id` can be `None` at that point, so `final_response_sent` is never set even though text was already shown.
2. **Cache-invalidation rebuild.** A mid-turn `Agent cache invalidated … message_count changed (N -> N+1), possible cross-process write` rebuilds the agent/consumer; the consumer inspected at suppression time has default flags. (Note: invalidation alone does **not** always cause it — turns where the flags survived suppressed correctly.)

Diagnostic capture across one evening (one thread), each line immediately followed by a normal `Sending response` → visible duplicate:

```
streamed=False content_delivered=False visible_matches_final=False message_id=None
  visible_len=476 final_len=474
streamed=False content_delivered=False visible_matches_final=False message_id=None
  visible_len=678 final_len=676
streamed=False content_delivered=False visible_matches_final=False message_id=None
  visible_len=430 final_len=485
streamed=False content_delivered=False visible_matches_final=False message_id=None
  visible_len=281 final_len=336
```

Note `visible_len` > 0 with all flags default: `_reset_segment_state()` clears `_last_sent_text` too, so a non-empty visible prefix means **no segment reset happened after the final text** — the flags were lost via the rebuild/cancel path, not a tool boundary.

### Suggested fix

Two complementary changes (both verified to stop the duplicates on our deployment):

1. **Tolerant `_visible_matches_final`.** Compare on a whitespace-normalized basis instead of exact equality, and accept prefix/superset overlap: suppress when normalized visible == final, or visible starts-with final (cleaning residue), or final starts-with visible with coverage ≥ ~0.80 (the coverage floor avoids silently truncating a genuinely unfinished stream).
2. **Carry delivery across rebuild.** A short-TTL (≈8s) per-session "final already delivered" ledger so a freshly-rebuilt consumer (state fully reset, no visible text) still suppresses its redundant send for the same turn. Gate it strictly (no visible text, no flags, normalized-final match, TTL) so a legitimate later turn can't be swallowed.

Alternatively/additionally: in the `CancelledError` handler, don't require `_message_id` to confirm delivery when `_last_sent_text` already covers the final answer.

### Related (checked, not duplicates)

#51010, #50965, #52095 (all >4096 overflow/split or flood-control retry paths), #53179 (llama.cpp raw+formatted render). This report is the non-overflow, flags-lost suppression path.

### PR DESCRIPTION
**Pull request by itsflownium**
Problem
Telegram streaming could show a complete short reply, lose the stream-consumer delivery flags, then send the same final response again. The fallback duplicate-suppression path only trusted strict flags and exact text equality, so cursor, whitespace, or formatting residue caused the normal final send to fire.

Summary
- Add normalized visible-vs-final comparison for streamed text.
- Treat exact normalized matches, visible supersets, and high-coverage visible prefixes as already delivered.
- Use that check for both queued-follow-up delivery and the final already_sent suppression path.
- Add regression tests for cursor/whitespace residue and high-vs-low coverage prefixes.

Validation
- $HOME/.hermes/hermes-agent/venv/bin/python -m pytest -q tests/gateway/test_duplicate_reply_suppression.py
- $HOME/.hermes/hermes-agent/venv/bin/ruff check gateway/run.py tests/gateway/test_duplicate_reply_suppression.py
- git diff --check

Fixes #53449

### CODE REVIEW

**Review by tonydwb — COMMENTED** [review]
## Code Review Summary

**Verdict: Comment**

Addresses the same duplicate reply suppression issue as PR #53516, with a different approach that includes a minimum coverage threshold (80%). Adds `_normalize_stream_delivery_text()` with cursor-suffix regex, `_stream_visible_matches_final()` with coverage ratio check, and `_stream_consumer_visible_text()` helper.

**Note:** Competing fix pair with PR #53516. This approach adds an 80% minimum coverage threshold (visible text must cover at least 80% of final text) while #53516 uses exact/prefix matching. The maintainer should reconcile which approach to merge.

---
*Reviewed by Hermes Agent*

File: gateway/run.py [hunk 1]
```diff
@@ -103,6 +103,41 @@ def _gateway_surface_passes_raw_text(platform: Any) -> bool:
...
+    """Return True when streamed visible text already covers final text."""
+    visible_norm = _normalize_stream_delivery_text(visible)
+    final_norm = _normalize_stream_delivery_text(final)
+    if not visible_norm or not final_norm:
+        return False
+    if visible_norm == final_norm or visible_norm.startswith(final_norm):
+        return True
+    if final_norm.startswith(visible_norm):
+        return len(visible_norm) / max(len(final_norm), 1) >= _STREAM_VISIBLE_FINAL_MIN_COVERAGE
+    return False
```
Comment (teknium1) [inline gateway/run.py:126]: This accepts an incomplete streamed prefix as final delivery at 80% coverage, so the caller suppresses the send that contains the missing suffix. Please require exact cleaned delivery or an explicit final-delivery signal; do not use a length threshold for user-visible content.

**Review by teknium1 — COMMENTED** [review]
Thanks for isolating the duplicate-delivery path and covering cursor residue.

### Problems
- `gateway/run.py:126` treats an 80%-complete visible prefix as final delivery. That can suppress the normal send while the remaining suffix was never delivered. This is particularly unsafe given the linked #53449 overflow report of missing trailing chunks.
- The new tests cover only the standalone matcher. Current main now keeps the final-delivery contract in `gateway/stream_consumer.py:313-321` and final-content state (`gateway/stream_consumer.py:1815-1825`); the proposed flow needs an end-to-end stream-consumer regression rather than matcher-only coverage.

### Suggested changes
- Do not suppress based on prefix coverage; require explicit final-delivery state or exact cleaned delivered text.
- Rework against the current delivery-confirmation path and add a lifecycle test for the claimed cancellation/cache-invalidation scenario.

Automated hermes-sweeper review.
